onnxruntime/core/codegen/target/tvm_schedule_builder.h

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "core/codegen/target/tvm_scheduler.h"
+#include "core/common/common.h"
+
+namespace onnxruntime {
+namespace tvm_codegen {
+
+// TVMScheduleBuilder contains all applicable TVM scheduler passes.
+// Scheduler passes are stored in multiple dispatchers
+// that check different conditions of a tvm::Tensor.
+
+// If a tvm::Tensor satisfies more than one TVM scheduler passes,
+// the first dispatched pass will be applied.
+
+class TVMScheduleBuilder {
+ public:
+  // TODO: add more parameter in consructor to support different target
+  TVMScheduleBuilder(const std::string& name);
+  ~TVMScheduleBuilder() = default;
+
+  void DumpAllSchedulers() const;
+
+  Status Evaluate(
+      const tvm::Tensor& tensor,
+      const Node* node,
+      CodeGenContext& ctx,
```

Comment:
>CodeGenContext& ctx [](start = 6, length = 19)

as we talked before, today if we use the generic IR building in mti, this CodeGenContext in this signature is useless, I have write another iteration to build the context. I think it will be better if we take a customized BuildContextFunc here.

Reply:
What told BuildContextFunc? 
Builder or OpIRCreator?
We tried to make builder and OpIRCreator stateless. 
CodeGenContext is passed as a state.

Reply:
It is useful to handle domain/opset. See codegen/target/generic/op_ir_creator/tensor/slice.cc.

---
In reply to: [294445998](https://github.com/microsoft/onnxruntime/pull/888#discussion_r294445998) [](ancestors = 294445998)

Reply:
yes, CodeGenContext is the state, but based on current implementation, different user has different state, CodeGenContext is just a handle. But if i use the generic ir builder in mti, it doesn't give me any chance to deal with the handle.

---
In reply to: [294552479](https://github.com/microsoft/onnxruntime/pull/888#discussion_r294552479) [](ancestors = 294552479)